packages/eslint-plugin/src/rules/thenable-in-promise-aggregators.ts

```diff
@@ -0,0 +1,176 @@
+import {
+  isPromiseConstructorLike,
+  isPromiseLike,
+  isTypeAnyType,
+  isTypeUnknownType,
+} from '@typescript-eslint/type-utils';
+import type { TSESTree } from '@typescript-eslint/utils';
+import { AST_NODE_TYPES } from '@typescript-eslint/utils';
+import * as tsutils from 'ts-api-utils';
+import * as ts from 'typescript';
+
+import { createRule, getParserServices } from '../util';
+
+export default createRule({
+  name: 'thenable-in-promise-aggregators',
+  meta: {
+    docs: {
+      description:
+        'Disallow passing non-Thenable values to promise aggregators',
+      recommended: 'strict',
+      requiresTypeChecking: true,
+    },
+    messages: {
+      inArray:
+        'Unexpected non-Thenable value in array passed to promise aggregator.',
+      arrayArg:
+        'Unexpected array of non-Thenable values passed to promise aggregator.',
+    },
+    schema: [],
+    type: 'problem',
+  },
+  defaultOptions: [],
+
+  create(context) {
+    const services = getParserServices(context);
+    const checker = services.program.getTypeChecker();
+
+    const aggregateFunctionNames = ['all', 'race', 'allSettled', 'any'];
+
+    function skipChainExpression<T extends TSESTree.Node>(
+      node: T,
+    ): T | TSESTree.ChainElement {
+      return node.type === AST_NODE_TYPES.ChainExpression
+        ? node.expression
+        : node;
+    }
+
+    function isPartiallyLikeType(
+      type: ts.Type,
+      predicate: (type: ts.Type) => boolean,
+    ): boolean {
+      if (isTypeAnyType(type) || isTypeUnknownType(type)) {
```

Comment:
The main goal for this lint rule for me is to catch cases where you can statically guarrantee there is a simpler way to call the aggregator that fulfills in the same amount of time, i.e., where it is very, very likely you accidentally provided the wrong value. For providing `unknown` to a promise aggregator, I might guard it with a type check, but I think it'd be perfectly reasonable not to:

```ts
// not looking to cause an error with something like this
async function doThingWithT(t: unknown) {
  let timedOut = false;
  const timeout = new Promise((resolve) => {
    timedOut = true;
    setTimeout(resolve, 100);
  });
  await Promise.race([t, timeout]);
  if (timedOut) {
    throw new Error('t is thenable and exceeded timeout');
  }
  const awaitedT = await t;
  if (typeof awaitedT === 'string') { 
    // etc etc
  }
}
```

```ts
// am looking to cause an error with something like this (type taken from production code)
type CancelablePromise<T> = { done: () => boolean, promise: Promise<T>, cancel: () => void };

const doThingWithGuard = async (a: Promise<number>, b: CancelablePromise<number>) => {
  try {
    await Promise.race([a, b]); // <--- error here, should be b.promise. What's written is useless
  } catch (e) {
    // this error handling isn't gonna work
  }

  if (b.done()) {
    // since the above is incorrect, this could still take time and could still error
    const aResult = await a;
    // ...
  }
  // ...
}
```

Reply:
That first use case seems pretty niche to me. In that case I'd think a type guard _would_ be the right thing to use. Maybe I'm being overly prescriptive. 😄 

Either way, here's an example with an `unknown` that blatantly doesn't need to be passed in [<a href="https://deploy-preview-8094--typescript-eslint.netlify.app/play/#ts=5.3.3&fileType=.ts&code=IYZwngdgxgBAZgV2gFwJYHsIwDboOYBC6yAFgBQBQMMeApsgGKoBOIyAXDGQJQwC8APhgAFZugC2qELQA8bZqgh4BAGioxpUTABMAssDAAjWpyQBrCOgDuENbwDe6rRDYwA2nBZsVG2s%2B0AuvwwwFbAqMgiYpLSAHTA2NhkbnSMXsg8Ppo6%2Bka0AdwA3BROmCDo2LSxuHhk9vDpWX46MAC%2BRRStQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6fAC0SYEMAjJLWa1i0APYBbSskS0eAc3nRE8nvjHRUGSImjjokcGAC%2BIY0A&tsconfig=&tokens=false" title="Playground link with logBoth function passing unknown to Promise.all">playground link</a>]:

```ts
async function logBoth(
  getFirst: () => Promise<string>,
  secondMaybe: unknown,
) {
  const [first, second] = await Promise.all([getFirst(), secondMaybe]);

  console.log({ first, second });
}
```

I think the rule should complain on this by default and have an option to not.

Reply:
cc @bradzacher per https://github.com/typescript-eslint/typescript-eslint/issues/8301#issuecomment-1910253980 - would you say the same logic applies here? That we should ignore `unknown`s?